internal/recursive_delete.go

```diff
@@ -1,68 +1,44 @@
 package internal
 
 import (
-	"log"
 	"os"
 	"path/filepath"
 )
 
-type fsNode struct {
-	Parent   *fsNode
-	Children map[string]*fsNode
-	HasFiles bool
-	Path     string
-}
-
 // Recurse to the leaf directories looking for empty directories to delete
 // and then work our way back up deleting everything that becomes empy during
 // the process.
-func deleteEmptyDirs(n *fsNode) error {
-	f, err := os.Open(n.Path)
+func deleteEmptyDirs(path string) (bool, error) {
+	node, err := os.Open(path)
 	if err != nil {
-		return err
+		return false, err
 	}
-	defer f.Close()
-
-	dirs, err := f.ReadDir(-1)
+	defer node.Close()
+	children, err := node.ReadDir(-1)
 	if err != nil {
-		return err
+		return false, err
 	}
-	for _, d := range dirs {
-		if d.IsDir() {
-			child := &fsNode{
-				Parent:   n,
-				Children: make(map[string]*fsNode),
-				HasFiles: false,
-				Path:     filepath.Join(n.Path, d.Name()),
-			}
-			n.Children[child.Path] = child
-			if err = deleteEmptyDirs(child); err != nil {
-				return err
+	isEmpty := true
+	for _, child := range children {
+		if child.IsDir() {
+			childPath := filepath.Join(path, child.Name())
+			empty, err := deleteEmptyDirs(childPath)
+			if err != nil {
+				return false, err
+			} else if !empty {
+				isEmpty = false
 			}
 		} else {
-			n.HasFiles = true
+			isEmpty = false
 		}
 	}
-
-	if n.Parent != nil && !n.HasFiles && len(n.Children) == 0 {
-		if err := os.Remove(n.Path); err != nil {
-			// We can only really warn here. People are going do do stuff like:
-			// docker run -v /run/secrets/foo:/foo
-			// which could prevent us from being able to delete the directory
-			log.Printf("Unable to remove empty directory: %s - %s", n.Path, err)
-		} else {
-			delete(n.Parent.Children, n.Path)
-		}
+	if isEmpty {
+		return true, os.Remove(path)
```

Comment:
This means that now the top-level `SecretsDir` can be removed if there are no configuration files left. Is this what you want to achieve?

Your previous implementation left the top-level `SecretsDir` intact. Also, if you move the `os.Remove` inside the loop over children (like I did) - it would be left intact as well.

Reply:
a963daba8c268ca361a48867e85966063ccef099